### PR TITLE
posix: implement PRIME for fd <-> handle conversion

### PR DESCRIPTION
**Pull request by no92**
This is rough code for making the PRIME ioctls work for sway. The code is not optimal and the naming is weird, but I couldn't come up with something better.

Posting as draft, as I don't expect this to survive review unchanged.

mlibc counterpart is managarm/mlibc#367.

### CODE REVIEW

File: core/drm/include/core/drm/core.hpp [hunk 1]
```diff
@@ -210,6 +211,8 @@ struct Device {
 	uint32_t getMinHeight();
 	uint32_t getMaxHeight();
 
+	helix::UniqueDescriptor _posixLane;
+	std::map<uint64_t, uint32_t> _fd_mappings;
```
Comment (64) [inline core/drm/include/core/drm/core.hpp:215]: This should be camel case.

Comment (no92) [reply core/drm/include/core/drm/core.hpp:215]: This should be removed as it is unused.

File: kernel/thor/generic/thor-internal/thread.hpp [hunk 2]
```diff
@@ -37,7 +37,18 @@ smarter::borrowed_ptr<Thread> getCurrentThread();
 
 struct ActiveHandle { };
 
-struct Thread final : smarter::crtp_counter<Thread, ActiveHandle>, ScheduleEntity {
+struct Credentials {
+	Credentials(void);
+	void allocateCredentials(void);
```
Comment (avdgrinten) [inline kernel/thor/generic/thor-internal/thread.hpp:42]: `void` as argument doesn't do anything in C++. Also: why is `allocateCredentials` its own function? Why should one call it over the constructor?

File: core/drm/src/core.cpp [hunk 3]
```diff
@@ -930,6 +963,30 @@ drm_core::BufferObject *drm_core::File::resolveHandle(uint32_t handle) {
 	return it->second.get();
 };
 
+void drm_core::File::exportBufferObject(uint32_t handle, std::array<char, 16> creds) {
+	auto bo = resolveHandle(handle);
+	assert(bo);
```
Comment (avdgrinten) [inline core/drm/src/core.cpp:968]: What if the handle is illegal?

File: core/drm/src/core.cpp [hunk 4]
```diff
@@ -930,6 +963,30 @@ drm_core::BufferObject *drm_core::File::resolveHandle(uint32_t handle) {
...
+	auto buffer = bo->sharedBufferObject();
+
+	_device->exportBufferObject(buffer, creds);
+	_primeBufferMap.insert({creds, handle});
+}
+
+std::pair<std::shared_ptr<drm_core::BufferObject>, uint32_t>
+drm_core::File::importBufferObject(std::array<char, 16> creds) {
+	auto bo = _device->importBufferObject(creds);
+	assert(bo);
```
Comment (avdgrinten) [inline core/drm/src/core.cpp:978]: What if the credentials are illegal?

File: core/drm/src/core.cpp [hunk 5]
```diff
@@ -930,6 +963,30 @@ drm_core::BufferObject *drm_core::File::resolveHandle(uint32_t handle) {
...
+
+	auto it = _primeBufferMap.find(creds);
+	if(it != _primeBufferMap.end()) {
+		return {bo, it->second};
+	} else {
+		uint32_t handle = createHandle(bo);
+		_primeBufferMap.insert({creds, handle});
+		return {bo, handle};
+	}
```
Comment (avdgrinten) [inline core/drm/src/core.cpp:987]: Why do we insert something in the `import` function? Isn't exportBufferObject supposed to do that?

File: core/drm/include/core/drm/core.hpp [hunk 6]
```diff
@@ -248,7 +250,12 @@ struct Device {
 	std::shared_ptr<Property> _crtcWProperty;
 	std::shared_ptr<Property> _crtcHProperty;
 
+	std::map<std::array<char, 16>, std::shared_ptr<drm_core::BufferObject>> _exportedBufferObjects;
+
 public:
+	void exportBufferObject(std::shared_ptr<drm_core::BufferObject> obj, std::array<char, 16> creds);
+	std::shared_ptr<drm_core::BufferObject> importBufferObject(std::array<char, 16> creds);
```
Comment (avdgrinten) [inline core/drm/include/core/drm/core.hpp:257]: I don't understand what the difference between these functions in `File` and `Device` is. It doesn't help that they have the same name.

File: core/drm/src/core.cpp [hunk 7]
```diff
@@ -202,6 +216,10 @@ drm_core::Device::Device() {
 		bool validate(const Assignment&) override {
 			return true;
 		};
+
+		void writeToState(const Assignment assignment, std::unique_ptr<AtomicState> &state) override {
+			state->crtc(assignment.object->id())->active = assignment.intValue;
+		}
```
Comment (avdgrinten) [inline core/drm/src/core.cpp:222]: Unrelated change? Seems that this function is never called.

File: core/drm/src/core.cpp [hunk 8]
```diff
@@ -1733,6 +1816,88 @@ drm_core::File::ioctl(void *object, managarm::fs::CntRequest req,
...
+		auto [creds] = co_await helix_ng::exchangeMsgs(conversation, helix_ng::extractCredentials());
+		HEL_CHECK(creds.error());
+
+		// 'import' the BufferObject while returning or creating the DRM handle that references it
+		std::array<char, 16> credentials;
+		std::copy_n(creds.credentials(), 16, std::begin(credentials));
+		auto [bo, handle] = self->importBufferObject(credentials);
+
+		assert(handle);
+		assert(bo);
```
Comment (avdgrinten) [inline core/drm/src/core.cpp:1890]: This is missing error handling.

**Review by avdgrinten — COMMENTED** [review]
Architecturally, the PR is good now. Some minor issues still need to be fixed.

File: core/drm/src/core.cpp (hunk 7 above)
Comment (no92) [reply core/drm/src/core.cpp:222]: While unused now, I have some patches to drivers that make this necessary as some piece of software relied on this property.

File: core/drm/src/core.cpp (hunk 3 above)
Comment (no92) [reply core/drm/src/core.cpp:968]: If the handle is invalid/unknown, `resolveHandle` returns a nullptr. The assert is supposed to catch that.

File: core/drm/include/core/drm/core.hpp (hunk 6 above)
Comment (no92) [reply core/drm/include/core/drm/core.hpp:257]: In `Device`, (credential, BO) pairs are registered so they can be imported back into `File`s. Each `File` has their own mappings for DRM handles, and that's what the functions in `File` set up. It might however be better to rename the functions in `Device` to something like `registerBufferObject` and `findBufferObject` to avoid confusion.

File: core/drm/src/core.cpp [hunk 9]
```diff
@@ -913,7 +942,11 @@ const std::vector<std::shared_ptr<drm_core::FrameBuffer>> &drm_core::File::getFr
 
 uint32_t drm_core::File::createHandle(std::shared_ptr<BufferObject> bo) {
 	auto handle = _allocator.allocate();
-	_buffers.insert({handle, bo});
+	auto ret = _buffers.insert({handle, bo});
+	assert(ret.second);
```
[the remark this answers is not in the document]
Comment (no92) [reply core/drm/src/core.cpp:946]: Hopefully not, but this is a sanity check to catch the (unlikely) case of duplicate handles.

File: core/drm/src/core.cpp (hunk 4 above)
Comment (no92) [reply core/drm/src/core.cpp:978]: If the credentials are illegal, the assert would trip due to `importBufferObject` returning a nullptr.

File: core/drm/src/core.cpp (hunk 5 above)
Comment (no92) [reply core/drm/src/core.cpp:987]: The `_primeBufferMap` is the `File`-local mapping between credentials and DRM handles, which can change between `File`s. This fact is the entire reason for PRIME using fds and not the pre-existing DRM handles. The `File::importBufferObject` function returns a pair of (BO, DRM handle) read from the `_primeBufferMap`, which is used in the PRIME ioctls.

File: core/drm/include/core/drm/core.hpp [hunk 10]
```diff
@@ -357,6 +369,8 @@ struct File {
 	std::unordered_map<uint32_t, std::shared_ptr<BufferObject>> _buffers;
 	// id allocator for mapping BufferObjects
 	id_allocator<uint32_t> _allocator;
+	// a mapping of credentials to DRM handles for PRIME
+	std::map<std::array<char, 16>, uint32_t> _primeBufferMap;
```
Comment (avdgrinten) [inline core/drm/include/core/drm/core.hpp:373]: Why doesn't this simply map the credentials to a shared_ptr that points to the buffer? The redirection over handles seems unnecessary (also handles can be deallocated and reallocated).

Comment (no92) [reply core/drm/include/core/drm/core.hpp:373]: We do not currently free ids in `DESTROY_DUMB` as it turns out, so handle reassignment does not happen with current code.

**Comment by no92** [comment]
Removed `_primeBufferMap`, squashed the fixups, dropped one commit as requested.

**Review by avdgrinten — APPROVED** [review]
LGTM, thank you for the PR!
